File: src/lib/vec3.ts
import { pipe, flow } from 'https://deno.land/x/hkts@v0.0.49/fns.ts'
import * as R from '../utils/random.ts'

export type Vec3 =
  { x: number; y: number; z: number }



// Constructors

export const vec3 = 
  (x: number, y: number, z: number) =>
  ({ x, y, z })

export const random =
  () =>
  vec3(R.random(), R.random(), R.random())

export const randomRange =
  (min: number) => (max: number) =>
  vec3(R.randomRange(min)(max), R.randomRange(min)(max), R.randomRange(min)(max))

export const randomInUnitSphere =
  () =>
  {
    while (true) {
      const p = randomRange(-1)(1)
      if (lengthSquared(p) < 1)
        return p
    }
  }

export const randomInUnitDisk =
  () => 
  {
    while (true) {
      const p = vec3(R.randomRange(-1)(1), R.randomRange(-1)(1), 0)
      if (lengthSquared(p) < 1)
        return p
    }
  }

export const randomUnitVector =
  () =>
  unitVector(randomInUnitSphere())




// Properties

export const lengthSquared =
  ({ x, y, z }: Vec3) => x*x + y*y + z*z

export const length =
  flow(lengthSquared, Math.sqrt)

export const nearZero =
  ({ x, y, z }: Vec3) =>
  {
    const s = 1e-8
    return (Math.abs(x) < s) && (Math.abs(y) < s) && (Math.abs(z) < s)
  }



// Operations

export const multiply =
  (s: number) => (v: Vec3) =>
  ({
    x: v.x * s,
    y: v.y * s,
    z: v.z * s
  })

export const divide =
  (s: number) =>
  multiply(1 / s)

export const negate =
  multiply(-1)

export const add =
  (b: Vec3) => (a: Vec3) =>
  ({
    x: a.x + b.x,
    y: a.y + b.y,
    z: a.z + b.z,
  })

export const subtract =
  (b: Vec3) => (a: Vec3) =>
  add(a)(negate(b))

export const unitVector =
  (v: Vec3) =>
  divide(length(v))(v)

export const dot =
  (v: Vec3) => (u: Vec3) =>
  u.x*v.x + u.y*v.y + u.z*v.z

export const cross =
  (v: Vec3) => (u: Vec3) =>
  ({
    x: u.y * v.z - u.z * v.y,
    y: u.z * v.x - u.x * v.z,
    z: u.x * v.y - u.y * v.x
  })

export const attenuate =
  (attenuation: Vec3) => (v: Vec3) =>
  ({
    x: attenuation.x * v.x,
    y: attenuation.y * v.y,
    z: attenuation.z * v.z
  })

export const reflect =
  (n: Vec3) => (v: Vec3) =>
  pipe(
    v,
    subtract(
      pipe(
        n,
        multiply(2 * dot(v)(n))
      )
    )
  )

export const refract =
  (uv: Vec3) => (n: Vec3) => (etaiOverEtat: number) =>
  {
    const cosTheta = Math.min(dot(negate(uv))(n), 1)
    const rOutPerp = multiply(etaiOverEtat)(add(uv)(multiply(cosTheta)(n)))
    const rOutParallel = multiply(-Math.sqrt(Math.abs(1 - lengthSquared(rOutPerp))))(n)
    return add(rOutPerp)(rOutParallel)
  }

File: src/vec3.rs
use crate::utils::{random_double, random_range};
use core::ops::{Add, Div, Mul, Sub};
use std::ops::Neg;

#[derive(Copy, Clone)]
pub struct Vec3(pub f64, pub f64, pub f64);

impl Vec3 {
    pub fn length_squared(self) -> f64 {
        self.0 * self.0 + self.1 * self.1 + self.2 * self.2
    }

    pub fn length(self) -> f64 {
        self.length_squared().sqrt()
    }

    pub fn unit_vector(self) -> Vec3 {
        self / self.length()
    }

    pub fn dot(self, other: Vec3) -> f64 {
        self.0 * other.0 + self.1 * other.1 + self.2 * other.2
    }

    pub fn reflect(self, n: Vec3) -> Vec3 {
        self - 2.0 * self.dot(n) * n
    }

    pub fn refract(self, n: Vec3, refraction_ratio: f64) -> Vec3 {
        let cos_theta = -self.dot(n).min(1.0);
        let r_out_perp = refraction_ratio * (self + cos_theta * n);
        let r_out_parallel = -(1.0 - r_out_perp.length_squared()).abs().sqrt() * n;
        r_out_perp + r_out_parallel
    }

    pub fn near_zero(self) -> bool {
        // Return true if the vector is close to zero in all dimensions
        let s = 1e-8;
        self.0.abs() < s && self.1.abs() < s && self.2.abs() < s
    }

    pub fn origin() -> Vec3 {
        Vec3(0.0, 0.0, 0.0)
    }

    pub fn random() -> Vec3 {
        Vec3(random_double(), random_double(), random_double())
    }

    pub fn random_range(min: f64, max: f64) -> Vec3 {
        Vec3(
            random_range(min, max),
            random_range(min, max),
            random_range(min, max),
        )
    }

    pub fn random_in_unit_sphere() -> Vec3 {
        loop {
            let p = Vec3::random_range(-1.0, 1.0);
            if p.length_squared() < 1.0 {
                return p;
            }
        }
    }

    pub fn random_unit_vector() -> Vec3 {
        Vec3::random_in_unit_sphere().unit_vector()
    }
}

impl Neg for Vec3 {
    type Output = Vec3;

    fn neg(self) -> Vec3 {
        Vec3(-self.0, -self.1, -self.2)
    }
}

impl Add for Vec3 {
    type Output = Vec3;

    fn add(self, other: Vec3) -> Vec3 {
        Vec3(self.0 + other.0, self.1 + other.1, self.2 + other.2)
    }
}

impl Sub for Vec3 {
    type Output = Vec3;

    fn sub(self, other: Vec3) -> Vec3 {
        self + (-other)
    }
}

impl Mul for Vec3 {
    type Output = Vec3;
    fn mul(self, other: Vec3) -> Vec3 {
        Vec3(self.0 * other.0, self.1 * other.1, self.2 * other.2)
    }
}

impl Mul<f64> for Vec3 {
    type Output = Vec3;

    fn mul(self, other: f64) -> Vec3 {
        Vec3(self.0 * other, self.1 * other, self.2 * other)
    }
}

impl Mul<Vec3> for f64 {
    type Output = Vec3;

    fn mul(self, other: Vec3) -> Vec3 {
        other * self
    }
}

impl Div<f64> for Vec3 {
    type Output = Vec3;

    fn div(self, other: f64) -> Vec3 {
        self * (1.0 / other)
    }
}
